### PR TITLE
Call hierarchy for TS

### PR DESCRIPTION
**Pull request by JanKoehnlein**
Signed-off-by: Jan Koehnlein <jan.koehnlein@typefox.io>

### CODE REVIEW

**Comment by svenefftinge** [comment]
We should fix 
 - https://github.com/theia-ide/theia-callhierarchy-extension/issues/9 and 
 - https://github.com/theia-ide/theia-callhierarchy-extension/issues/8

**Comment by AlexTugarev** [comment]
I‘d like to look into this shortly. Currently trying to improve tests and will give an update on this.

File: packages/callhierarchy/package.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "@theia/callhierarchy",
+  "version": "0.3.2",
+  "description": "Theia - Call Hierarchy Extension",
+  "dependencies": {
+    "@theia/core": "^0.3.2",
+    "@theia/languages": "^0.3.2",
+    "@theia/editor": "^0.3.2",
+    "vscode-base-languageclient": "^0.0.1-alpha.3",
```
Comment (akosyakov) [inline packages/callhierarchy/package.json]: please remove this dependency, it is consumed via `@theia/core`

File: packages/callhierarchy/package.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    "vscode-languageserver-types": "^3.4.0",
```
Comment (akosyakov) [inline packages/callhierarchy/package.json]: please remove this dependency, it is consumed via `@theia/core`

File: packages/callhierarchy/src/browser/language-client-provider-impl.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+import { injectable, inject, named } from "inversify";
+import { ContributionProvider } from '@theia/core/lib/common';
+import { LanguageClientContribution } from '@theia/languages/lib/browser/language-client-contribution';
+import { ILanguageClient } from '@theia/languages/lib/common/languageclient-services';
+import { LanguageClientProvider } from './language-client-provider';
+
+@injectable()
+export class LanguageClientProviderImpl implements LanguageClientProvider {
```
Comment (akosyakov) [inline packages/callhierarchy/src/browser/language-client-provider-impl.ts]: We could move it to `@theia/languages`.

File: packages/java/src/node/java-contribution.ts [hunk 4]
```diff
@@ -35,6 +35,9 @@ export class JavaContribution extends BaseLanguageServerContribution {
         const jarPath = path.resolve(serverPath, jarPaths[0]);
         const workspacePath = path.resolve(os.tmpdir(), '_ws_' + new Date().getTime());
         const configuration = configurations.get(process.platform);
+        if (!configuration) {
+            throw new Error('Cannot find Java sevber configuration for ' + process.platform);
```
Comment (AlexTugarev) [inline packages/java/src/node/java-contribution.ts]: typo

File: packages/callhierarchy/README.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Theia - Call Hierarchy Extension
+
+Displays the caller hierarchy of a selected method.
```
Comment (AlexTugarev) [inline packages/callhierarchy/README.md]: ~method~ callable

File: packages/callhierarchy/src/browser/callhierarchy-service-impl.ts [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+    protected async getCallableSymbolsInResource(client: ILanguageClient, uri: string): Promise<SymbolInformation[]> {
+        const symbols = await client.sendRequest(DocumentSymbolRequest.type, <DocumentSymbolParams>{
+            textDocument: TextDocumentIdentifier.create(uri)
+        });
+        return symbols.filter(info => this.isCallable(info.kind));
+    }
+
+    /**
+     * Returns all method symbols from resources of given refences.
```
Comment (AlexTugarev) [inline packages/callhierarchy/src/browser/callhierarchy-service-impl.ts]: ~method~callable

**Comment by akosyakov** [comment]
please rebase this PR

File: packages/callhierarchy/src/browser/callhierarchy-service-impl.ts [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    dispose() {
+        this.disposables.forEach(element => {
+            element.dispose();
+        });
+        this.symbolCache.clear();
+    }
+}
+
+@injectable()
+export abstract class CallHierarchyServiceImpl implements CallHierarchyService {
```
Comment (svenefftinge) [inline packages/callhierarchy/src/browser/callhierarchy-service-impl.ts]: Would be good to give it a name that indicates its best-effort implementation. Also `Abstract` prefix would be good.

File: packages/callhierarchy/src/browser/active-editor-access.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    getLanguageId(): string | undefined {
+        const activeEditor = this.getActiveEditor();
+        if (!activeEditor) {
+            return;
+        }
+        return activeEditor.document.languageId;
+    }
+
+    protected getActiveEditor(): TextEditor | undefined {
+        const activeEditor = this.editorManager.currentEditor;
```
[the remark this answers is not in the document]
Comment (JanKoehnlein) [reply packages/callhierarchy/src/browser/active-editor-access.ts]: Changed to _active_.

File: packages/callhierarchy/src/browser/callhierarchy-contribution.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            defaultWidgetOptions: {
+                area: 'bottom'
+            },
+            toggleCommandId: CALL_HIERARCHY_TOGGLE_COMMAND_ID,
+            toggleKeybinding: 'ctrlcmd+shift+f1'
+        });
+    }
+
+    initializeLayout(app: FrontendApplication) {
+        // TODO add status bar item
```
[the remark this answers is not in the document]
Comment (JanKoehnlein) [reply packages/callhierarchy/src/browser/callhierarchy-contribution.ts]: Dunno, was there from Alex's initial contrib.

File: packages/callhierarchy/src/browser/callhierarchy-contribution.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    protected isCallHierarchyAvailable(): boolean {
+        const selection = this.editorAccess.getSelection();
+        const languageId = this.editorAccess.getLanguageId();
+        return !!selection && !!languageId;
```
[the remark this answers is not in the document]
Comment (JanKoehnlein) [reply packages/callhierarchy/src/browser/callhierarchy-contribution.ts]: Check added.

File: packages/callhierarchy/src/browser/callhierarchy-service-impl.ts (hunk 7 above)
Comment (JanKoehnlein) [reply packages/callhierarchy/src/browser/callhierarchy-service-impl.ts]: Changed to `AbstractDefaultCallhierarchyService`

**Comment by svenefftinge** [comment]
Could you take care of the compile changes https://travis-ci.org/theia-ide/theia/jobs/340094190#L894

**Comment by JanKoehnlein** [comment]
Rebased and adapted to API changes

**Comment by svenefftinge** [comment]
I think the 'current' vs. 'active' change makes it so that right clicking in the editor disables the command in the context menu when you move the mouse. (Hard to explain, please try :-))

**Comment by JanKoehnlein** [comment]
You are right. I switched to current editor

**Pull request merged**